states/visited2.go

```diff
@@ -0,0 +1,64 @@
+package states
+
+import (
+	"context"
+	"sync"
+
+	"github.com/earthly/earthly/domain"
+	"github.com/earthly/earthly/util/platutil"
+	"github.com/earthly/earthly/variables"
+)
+
+// VisitedCollections is a collection of visited targets.
+type VisitedCollection2 struct {
+	mu      sync.Mutex
+	visited map[string]*SingleTarget // targetInputHash -> sts
+	// Same collection as above, but as a list, to make the ordering consistent.
+	visitedList []*SingleTarget
+}
+
+// NewVisitedCollection returns a collection of visited targets.
+func NewVisitedCollection2() *VisitedCollection2 {
+	return &VisitedCollection2{
+		visited: make(map[string]*SingleTarget),
+	}
+}
+
+// All returns all visited items.
+func (vc *VisitedCollection2) All() []*SingleTarget {
+	vc.mu.Lock()
+	defer vc.mu.Unlock()
+	return append([]*SingleTarget{}, vc.visitedList...)
+}
+
+// Add adds a target to the collection, if it hasn't yet been visited. The returned sts is
+// either the previously visited one or a brand new one.
+// This function blocks if there is a visited target that is still running.
```

Comment:
>// This function blocks if there is a visited target that is still running.

This doesn't seem accurate, since it only blocks if there is cache hit to a running target?

Reply:
Yes, correct - it only blocks if the target input hash matches.

Reply:
What is the rationale behind this behavior?  
I mean why the first time visited it doesn't block but successive calls do?

Reply:
I ended up moving it out, so that I can use the hash for infinite recursion checking before waiting.